### PR TITLE
webgpu: support MultiHeadAttention operator

### PR DESCRIPTION
**Pull request by xhcao**

### CODE REVIEW

**Comment by xhcao** [comment]
Although I think there are some unreasonable codes in JS MultiHeadAttention operator, I still kept the code in webgpu EP nearly the same as JS EP. Let's adjust and optimize the code in future.
@fs-eire @qjia7

File: onnxruntime/contrib_ops/webgpu/bert/multihead_attention.h [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+
+namespace onnxruntime {
+namespace contrib {
+namespace webgpu {
+
+using namespace onnxruntime::webgpu;
+
+class TransferBSDToBNSHProgram final : public Program<TransferBSDToBNSHProgram> {
+ public:
+  TransferBSDToBNSHProgram(const std::string& kernel_name, bool has_bias) : Program{kernel_name}, has_bias_(has_bias) {}
```
Comment (fs-eire) [inline onnxruntime/contrib_ops/webgpu/bert/multihead_attention.h:19]:
```suggestion
  TransferBSDToBNSHProgram(bool has_bias) : Program{"TransferBSDToBNSH"}, has_bias_(has_bias) {}
```

File: onnxruntime/contrib_ops/webgpu/bert/multihead_attention.cc [hunk 2]
```diff
@@ -0,0 +1,506 @@
...
+  if (has_attention_bias_) {
+    shader.AddInput("attention_bias", ShaderVariable::UseUniform);
+  }
+
+  shader.AddOutput("output", ShaderVariable::UseUniform | ShaderVariable::UseValueTypeAlias);
+  if (has_present_key_) {
+    shader.AddOutput("present_key", ShaderVariable::UseUniform);
+  }
+
+  shader.AppendImplementation("const TILE_SIZE = ", tile_size_, "u;\n")
```
Comment (fs-eire) [inline onnxruntime/contrib_ops/webgpu/bert/multihead_attention.cc:94]: it seems that cache keys for the programs are not set correctly.

for example, here `tile_size_` is used as a part of the shader source code, but it is not set in the cache key. Use `program.CacheHint()` to set the cache key.

Comment (fs-eire) [reply onnxruntime/contrib_ops/webgpu/bert/multihead_attention.cc:94]: TILE_SIZE can also be declared in overridable constants.

**Comment by fs-eire** [comment]
I didn't see any call to set program cache key. This may be correct (if necessary information is already in uniform). need to confirm.

File: onnxruntime/contrib_ops/webgpu/bert/multihead_attention.cc (hunk 2 above)
Comment (xhcao) [reply onnxruntime/contrib_ops/webgpu/bert/multihead_attention.cc]: Done

File: onnxruntime/contrib_ops/webgpu/bert/multihead_attention.cc [hunk 3]
```diff
@@ -0,0 +1,506 @@
...
+  return context.RunProgram(program);
+}
+
+Status InPlaceSoftmaxProgram::GenerateShaderCode(ShaderHelper& shader) const {
+  shader.AddOutput("x", ShaderVariable::UseUniform | ShaderVariable::UseValueTypeAlias |
+                        ShaderVariable::UseElementTypeAlias);
+  shader.AppendImplementation("var<workgroup> thread_max: array<f32, ", work_group_size_, ">;\n")
+        .AppendImplementation("var<workgroup> thread_sum: array<f32, ", work_group_size_, ">;\n");
+
+  std::string f32_str = components_ == 4 ? "vec4<f32>" : (components_ == 2 ? "vec2<f32>" : "f32");
```
[the remark this answers is not in the document]
Comment (xhcao) [reply onnxruntime/contrib_ops/webgpu/bert/multihead_attention.cc:214]: No matter `x`'s type is f32 or f16, the program only uses f32 to define max and sum values.

File: onnxruntime/contrib_ops/webgpu/bert/multihead_attention.h (hunk 1 above)
Comment (xhcao) [reply onnxruntime/contrib_ops/webgpu/bert/multihead_attention.h]: Done

**Comment by fs-eire** [comment]
Merged to latest.

**Pull request merged**